badger/cmd/info.go

```diff
@@ -59,39 +63,78 @@ info. It also prints info about missing/extra files, and general information abo
 files (which are not referenced by the manifest).  Use this tool to report any issues about Badger
 to the Dgraph team.
 `,
-	Run: func(cmd *cobra.Command, args []string) {
-		err := printInfo(sstDir, vlogDir)
-		if err != nil {
-			fmt.Println("Error:", err.Error())
-			os.Exit(1)
-		}
-		if !opt.showTables {
-			return
-		}
-		// Open DB
-		opts := badger.DefaultOptions
-		opts.TableLoadingMode = options.MemoryMap
-		opts.Dir = sstDir
-		opts.ValueDir = vlogDir
-		opts.ReadOnly = true
-
-		db, err := badger.Open(opts)
-		if err != nil {
-			fmt.Println("Error:", err.Error())
-			os.Exit(1)
-		}
-		defer db.Close()
+	RunE: handleInfo,
+}
+
+func handleInfo(cmd *cobra.Command, args []string) error {
+	err := printInfo(sstDir, vlogDir)
+	if err != nil {
+		fmt.Println("Error:", err.Error())
+		os.Exit(1)
+	}
+
+	// Open DB
+	opts := badger.DefaultOptions
+	opts.TableLoadingMode = options.MemoryMap
+	opts.Dir = sstDir
+	opts.ValueDir = vlogDir
+	opts.ReadOnly = true
+
+	db, err := badger.Open(opts)
+	if err != nil {
+		fmt.Println("Error:", err.Error())
+		os.Exit(1)
+	}
+	defer db.Close()
 
+	if opt.showTables {
 		err = tableInfo(sstDir, vlogDir, db)
 		if err != nil {
 			fmt.Println("Error:", err.Error())
 			os.Exit(1)
 		}
-		if opt.sizeHistogram {
-			// use prefix as nil since we want to list all keys
-			db.PrintHistogram(nil)
+	}
+
+	if opt.sizeHistogram {
+		db.PrintHistogram([]byte(opt.withPrefix))
+	}
+
+	if opt.dumpData {
+		if err := dumpData(db, opt.withPrefix); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func dumpData(db *badger.DB, prefix string) error {
+	if len(prefix) > 0 {
+		fmt.Printf("Only choosing keys with prefix: %s\n", prefix)
+	}
+	fmt.Printf("\nInformation about keys and values in the store\n")
+	err := db.View(func(txn *badger.Txn) error {
+		it := txn.NewIterator(badger.DefaultIteratorOptions)
+		defer it.Close()
+		totalKeys := 0
+		for it.Seek([]byte(prefix)); it.ValidForPrefix([]byte(prefix)); it.Next() {
```

Comment:
do the conversion just once from string to bytes.

Reply:
Done.